### ISSUE TITLE
nodeSets not working in parallel

### ISSUE THREAD

**Issue by plovely** [underworldcode/UWGeodynamics#111]
I am using nodeSets to define a thermal anomaly to seed deformation.  It appears the node numbers are assigned in each parallel domain rather than using global indices.  

Image of initial thermal state from analysis on 4 CPU's:
![image](https://user-images.githubusercontent.com/30320218/55423841-7294c400-5544-11e9-84ef-30d1613dab40.png)

Code used to define thermal boundary conditions:
![image](https://user-images.githubusercontent.com/30320218/55423892-8f30fc00-5544-11e9-9a6f-13cf4c3f12ed.png)

The same result occurs using docker images tagged "latest" or "v2.7.1"

**Comment by rbeucher** [comment]
Hi Peter,

Yes that's the expected behaviour because you are passing a range of indices. 
What are you trying to do?

**Comment by plovely** [comment]
Hi Romain,
I am trying to define a thermal perturbation (i.e. elevated temperature) on the 30 nodes at the center of the model base. 
Thanks,
Peter

**Comment by plovely** [comment]
This is what the steady-state thermal field should look like, with just one anomaly in the middle:
![image](https://user-images.githubusercontent.com/30320218/55429673-9f030d00-5551-11e9-944e-c7bd6f770dd4.png)

**Comment by rbeucher** [comment]
OK so the best way to do that is to use a function F(x,y) with x,y being the coordinates.  You can also use a shape and that's probably the easiest in your case.
The Hot Canon Ball example is a good illustration. In your case you could use a box and it would work no matter the resolution.
https://github.com/underworldcode/UWGeodynamics/blob/master/examples/1_25_Hot_Canon_Ball.ipynb

The nodeSet parameter used to be able to handle global indices (a while ago..) but this quickly falls appart with 3D models and it implies some knowledge of the mesh structure..

**Comment by plovely** [comment]
Thanks, Romain.  It's working now.

**Comment by plovely** [comment]
This is a better solution.  I like it.

**Comment by CHENMING27** [comment]
I have a similiar question! Now, I am using UWGeodynamics to do researches about mid-ocean ridge evolution and I will apply an internal velocity boundary condition.  How can I apply this  kinematics boundary condition?

Thanks!

**Comment by CHENMING27** [comment]
The following is my code, it does not work.

npoints = 100

y1 = np.linspace(GEO.nd(-16. * u.kilometer), GEO.nd(0. * u.kilometer), npoints)
AxisNorthBoundary = np.ndarray((npoints, 2))

AxisNorthBoundary[:, 0] = 49.5
AxisNorthBoundary[:, 1] = y1


cxpos = np.ones(5)*49.5
cypos = np.linspace(GEO.nd(-16*u.kilometer), GEO.nd(0*u.kilometer),5)

cxpos, cypos = np.meshgrid(cxpos, cypos)
centroid_coords = np.ndarray((len(cxpos.ravel()), 2))
centroid_coords[:, 0] = cxpos.ravel()
centroid_coords[:, 1] = cypos.ravel()

Model.set_velocityBCs(left=[-7.1 * u.millimeter / u.year, None],
                       right=[8.0 * u.millimeter / u.year, None],
                       bottom=GEO.LecodeIsostasy(reference_mat=AsthenosphereMantle, average=True),
                       nodeSets=[(centroid_coords,-4.0 * u.millimeter / u.year)])

**Comment by julesghub** [comment]
Hi @CHENMING27,
Looks like `centroid_coords` is not a valid type for the nodeSets argument.

The docstring for the `set_velocityBCs()` says:
```
nodeSets: list of tuples: [(nodes, condition)]
            List of node where to apply predefined condition.
            The nodes can be a list or a numpy array containing the
            local index of the nodes. It can also be an Underworld
            IndexSet. You can also pass an UWGeodynamics shape.
            The condition can be an Underworld Function, A Pint
            Quantity of a scalar.
```
I would try the approach @rbeucher mentions above, i.e. using a `shape` to find the indices of the nodes.

---

To continue discussing this issue can you please open a separate ticket and please mention the version of the UWGeo you're using. Thanks!

**Comment by rbeucher** [comment]
Hi @CHENMING27 

Thank you for posting a question.
I will be more than happy to help but could you please open a new issue and explain what you are trying to achieve?

That makes things a lot easier for us

Romain

**Comment by CHENMING27** [comment]
Hi @rbeucher 

    My major is Marine Geology and I want to use UWGeodynamics code to make researches about evolution process of  mid-ocean ridge. As shown in the following figure, besides velocity boundary conditions (v1) applied on the left and right side of the model,  I also want to apply thses on the  left and right side of the axis(v2). But I do not know how to apply this internal velocity boundary conditions.
    My UWGeodynamics version is 2.10.1-dev-164585a(master). I use it in docker container.
    Thanks!
![Uploading fig1.png…]()

**Comment by CHENMING27** [comment]
![fig1](https://user-images.githubusercontent.com/72907912/96674432-5f9e4a80-139b-11eb-9d87-fa0441c19e99.png)

**Comment by CHENMING27** [comment]
Hi @julesghub 
     
   Thanks for your help!
